### DISCUSSION TITLE
How do I handle such a replacement macro using for default values of some parameters?

### DISCUSSION THREAD

**Discussion by lancewoo** [bytedeco/javacpp#563] [Q&A] [answered]
Hi @saudet , I am doing a translation of [mp4v2 ](https://github.com/TechSmith/mp4v2) on Android. 
Right now I am stuck in the macro definition
 [#   define DEFAULT(x) =x](https://github.com/TechSmith/mp4v2/blob/bfb06df99e55801650ac5bae287f10eb3e230111/include/mp4v2/mp4v2.h#L46). 
This macro is used everywhere in the interfaces of this library, for default values of some parameters, like this one:

```
MP4V2_EXPORT
MP4FileHandle MP4CreateEx(
    const char* fileName,
    uint32_t    flags DEFAULT(0),
    int         add_ftyp DEFAULT(1),
    int         add_iods DEFAULT(1),
    char*       majorBrand DEFAULT(0),
    uint32_t    minorVersion DEFAULT(0),
    char**      compatibleBrands DEFAULT(0),
    uint32_t    compatibleBrandsCount DEFAULT(0) );
```

I just used 

                infoMap.put(new Info("DEFAULT").cppTypes().annotations().cppText(""))

to skip it. Is that correct?

My whole map function goes like this:

    public void map(InfoMap infoMap) {
        infoMap
                .put(new Info("MP4_INVALID_FILE_HANDLE").cppTypes("MP4FileHandle"))
                .put(new Info("MP4V2_EXPORT").cppTypes().annotations().cppText(""))
                .put(new Info("DEFAULT").cppTypes().annotations().cppText(""))
                .put(new Info("defined( _WIN32 )").define(false))
                .put(new Info("defined( __MINGW32__ )").define(false))
                .put(new Info("defined( _WIN32 ) || defined( __MINGW32__ )").define(false))
                // https://github.com/bytedeco/javacpp-presets/blob/4c446f1e5016a3906551fc8ccf523c8dddd4ece4/libffi/src/main/java/org/bytedeco/libffi/presets/ffi.java#L80
                .put(new Info("__attribute__((deprecated))").annotations("@Deprecated"))
                // Of course we have __cplusplus for Android NDK bundle
                .put(new Info("__cplusplus").define(true))
                .put(new Info("defined( __cplusplus )").define(true))
                .put(new Info("!defined( __cplusplus )").define(false))
                // _MSC_VER should not be valid on Android
                .put(new Info("_MSC_VER").define(false));
    }

It seems the Parser and Compiler Gradle tasks work well now. I simply don't know if it works for this translation.
Thanks!

Lance

**Comment by saudet** [comment, marked as answer]
We shouldn't need to do anything special for default arguments, they get mostly ignored by default, but if you want to redefine a macro, what we need to do is shown on this page: https://github.com/bytedeco/javacpp/wiki/Mapping-Recipes#redefining-the-code-of-a-macro

Specifically in this case:
```java
infoMap.put(new Info("DEFAULT").cppText("#define DEFAULT(x)"));
```

**Reply by lancewoo** [reply]
It works now, thanks!

I also removed this macro translation:
```
#   define MP4V2_EXPORT __attribute__((visibility("default")))
```
and removed the include of mp4v2.h.

**Reply by lancewoo** [reply]
I just found if I don't skip the macro 
```#   define MP4V2_EXPORT __attribute__((visibility("default")))```
 with 
```infoMap.put(new Info("MP4V2_EXPORT").cppText("#define MP4V2_EXPORT"))```
 , return types of interfaces would not be translated correctly since all of them are prefixed with MP4V2_EXPORT. 

Is this a bug of the JavaCPP Gradle plugin or is this the duty for us developers to ensure its correct behavior?

@saudet

**Reply by lancewoo** [reply]
Oh, this has been explained on this page:
[Ignoring attributes and macros](https://github.com/bytedeco/javacpp/wiki/Mapping-Recipes#ignoring-attributes-and-macros)
Sorry I missed it.
Thanks!

**Comment by lancewoo** [comment]
I tried your approach. I seemed that didn't work.
Like this code:
```
/* 3GP specific utilities */

MP4V2_EXPORT
bool MP4Make3GPCompliant(
    const char* fileName,
    char*       majorBrand DEFAULT(0),
    uint32_t    minorVersion DEFAULT(0),
    char**      supportedBrands DEFAULT(NULL),
    uint32_t    supportedBrandsCount DEFAULT(0),
    bool        deleteIodsAtom DEFAULT(true) );
```

got translated as:

```
/* 3GP specific utilities */

public static native @Cast("bool") boolean MP4Make3GPCompliant(
    @Cast("const char*") BytePointer fileName,
    @Cast("char*") BytePointer majorBrand/*0*/,
    @Cast("uint32_t") int minorVersion/*0*/,
    @Cast("char**") PointerPointer supportedBrands/*NULL*/,
    @Cast("uint32_t") int supportedBrandsCount/*0*/,
    @Cast("bool") boolean true );
public static native @Cast("bool") boolean MP4Make3GPCompliant(
    @Cast("const char*") BytePointer fileName,
    @Cast("bool") boolean true );
public static native @Cast("bool") boolean MP4Make3GPCompliant(
    @Cast("const char*") BytePointer fileName,
    @Cast("char*") BytePointer majorBrand/*0*/,
    @Cast("uint32_t") int minorVersion/*0*/,
    @Cast("char**") @ByPtrPtr BytePointer supportedBrands/*NULL*/,
    @Cast("uint32_t") int supportedBrandsCount/*0*/,
    @Cast("bool") boolean true );
public static native @Cast("bool") boolean MP4Make3GPCompliant(
    String fileName,
    @Cast("char*") ByteBuffer majorBrand/*0*/,
    @Cast("uint32_t") int minorVersion/*0*/,
    @Cast("char**") @ByPtrPtr ByteBuffer supportedBrands/*NULL*/,
    @Cast("uint32_t") int supportedBrandsCount/*0*/,
    @Cast("bool") boolean true );
public static native @Cast("bool") boolean MP4Make3GPCompliant(
    String fileName,
    @Cast("bool") boolean true );
public static native @Cast("bool") boolean MP4Make3GPCompliant(
    @Cast("const char*") BytePointer fileName,
    @Cast("char*") byte[] majorBrand/*0*/,
    @Cast("uint32_t") int minorVersion/*0*/,
    @Cast("char**") @ByPtrPtr byte[] supportedBrands/*NULL*/,
    @Cast("uint32_t") int supportedBrandsCount/*0*/,
    @Cast("bool") boolean true );
public static native @Cast("bool") boolean MP4Make3GPCompliant(
    String fileName,
    @Cast("char*") BytePointer majorBrand/*0*/,
    @Cast("uint32_t") int minorVersion/*0*/,
    @Cast("char**") @ByPtrPtr BytePointer supportedBrands/*NULL*/,
    @Cast("uint32_t") int supportedBrandsCount/*0*/,
    @Cast("bool") boolean true );
public static native @Cast("bool") boolean MP4Make3GPCompliant(
    @Cast("const char*") BytePointer fileName,
    @Cast("char*") ByteBuffer majorBrand/*0*/,
    @Cast("uint32_t") int minorVersion/*0*/,
    @Cast("char**") @ByPtrPtr ByteBuffer supportedBrands/*NULL*/,
    @Cast("uint32_t") int supportedBrandsCount/*0*/,
    @Cast("bool") boolean true );
public static native @Cast("bool") boolean MP4Make3GPCompliant(
    String fileName,
    @Cast("char*") byte[] majorBrand/*0*/,
    @Cast("uint32_t") int minorVersion/*0*/,
    @Cast("char**") @ByPtrPtr byte[] supportedBrands/*NULL*/,
    @Cast("uint32_t") int supportedBrandsCount/*0*/,
    @Cast("bool") boolean true );

```

Anything wrong?

**Reply by saudet** [reply]
Sorry, just remove the x like this:
```java
infoMap.put(new Info("DEFAULT").cppText("#define DEFAULT(x)"));
```
I've updated the answer.
